File: 高级数据结构与算法/Python/第4章 模式匹配/4.3 RK/rk.py
def hash(s):
    hash_code = 0
    for c in s:
        hash_code += ord(c) - ord('a')
    return hash_code

def next_hash(s, hash, start, n):
    hash -= ord(s[start]) - ord('a')
    hash += ord(s[start + n]) - ord('a')
    return hash

def match(s, p):
    return s == p

def rk(s, p):
    s_len = len(s)
    p_len = len(p)

    s_hash = hash(s[0:p_len])       # 文本串子串哈希值
    p_hash = hash(p)                # 模式串哈希值

    for i in range(s_len - p_len + 1):
        if s_hash == p_hash:
            if match(s[i:i+p_len], p):
                return i
        
        if i < s_len - p_len:
            s_hash = next_hash(s, s_hash, i, p_len)

    return -1

def main():
    s = "a simple example"
    p = "ample"
    print(rk(s, p))

if __name__ == "__main__":
    main()

File: old_version/C/第4章 模式匹配/4.3 RK/rk.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

int hash(char *s, int n) {
    int hashCode = 0;
    for(int i = 0; i < n; i++) {
        hashCode += s[i] - 'a';
    }
    return hashCode;
}

int nextHash(char *s, int hash, int start, int n) {
    hash -= s[start] - 'a';
    hash += s[start + n] - 'a';
    return hash;
}

bool match(char *s, char *p, int n) {
    for(int i = 0; i < n; i++) {
        if(s[i] != p[i]) {
            return false;
        }
    }
    return true;
}

int rk(char *s, char *p) {
    int sLen = strlen(s);
    int pLen = strlen(p);

    int sHash = hash(s, pLen);  // 文本串子串哈希值
    int pHash = hash(p, pLen);  // 模式串哈希值

    for(int i = 0; i < sLen - pLen + 1; i++) {
        if(sHash == pHash) {
            if(match(s+i, p, pLen)) {
                return i;
            }
        }

        if(i < sLen - pLen) {
            sHash = nextHash(s, sHash, i, pLen);
        }
    }

    return -1;
}

int main() {
    char *s = "a simple example";
    char *p = "ample";
    printf("%d\n", rk(s, p));
    return 0;
}
